### PR TITLE
feat: program registry

### LINKED ISSUE #148: `Program` deployment and associated `register_program` contract registry predicate.
Once the new program `Predicate` type is ready to land with the updated `Program` DAG model, the contract registry should be updated in the node to allow `Solution`s to register `Program`s independently of `Predicate`s.

`Program`s can be stored using a unique key within the existing contract registry.

`Program`s will need to be deployed either in the same `Solution` that their dependent `Predicate`(s) are deployed in, or in any earlier `Solution`.

The contract registry mutation for deploying a program will likely look something like this:

```yaml
key: [2, <program-ca>]
value: [len_bytes, <program-bytecode-padded>]
```

where `2` is the `programs` storage prefix within the contract registry.

## Validation

Program deployment validation will involve ensuring:

1. The content address that appears in the key matches the actual program content address (i.e. `sha256` of the program bytecode).
2. The length in bytes does not exceed `Program::MAX_SIZE`.

### PR DESCRIPTION
**Pull request by supiket**
## Program registry
- [x] Program registry contract and program deployment
- [x] Tests

Like the contract registry contract, the program registry contract only has storage and no logic in the predicate for now. Related and next steps: #48, https://github.com/essential-contributions/essential-base/issues/190 and https://github.com/essential-contributions/essential-base/issues/221.

#### Review tip
Main change is to `BigBang` in `node-types/src/lib.rs`. Registry contract's layout is described here. In turn, function `validate_inner` in `crates/node/src/validate.rs` validates the programs.

## Update to latest base 
That means adapting to three major changes: DAG model of programs, https://github.com/essential-contributions/essential-base/pull/229 and https://github.com/essential-contributions/essential-base/pull/236.

#### Review tip 
The changes to `node-db-sql` and `node-db` are related solely to this. 

Closes #148

### CODE REVIEW

File: crates/node-types/src/lib.rs [hunk 1]
```diff
@@ -197,6 +232,45 @@ pub fn register_contract_mutations(contract: &Contract) -> Result<Vec<Mutation>,
...
+/// Learn more about the layout of state within the contract registry in
+/// [`BigBang::program_registry`].
+pub fn register_program_mutations(program: &Program) -> Vec<Mutation> {
+    let mut muts = vec![];
+
+    // Register the program.
+    let prog_ca = essential_hash::content_addr(program);
+    let prog_bytes = &program.0;
+    let len_bytes = prog_bytes.len();
+    let len_bytes_w = Word::try_from(len_bytes).expect("enforced by Program::MAX_SIZE");
```
Comment (supiket) [inline crates/node-types/src/lib.rs:259]: - [ ] Actually it is not but it should be

**Review by mitchmindtree — COMMENTED** [review]
AWESOME work! Totally nailed it :mage_woman: 

It looks like we might need to `cargo update -p essential-check` to update the `essential-base` crates in the Cargo.lock file and pull in the latest changes including `Solution` naming etc.

Otherwise lgtm!

File: crates/node-types/src/lib.rs [hunk 2]
```diff
@@ -61,6 +61,22 @@ pub struct BigBang {
     /// - `[1, <predicate-ca>]` to get the length bytes as `Word` followed by the fully encoded
     ///   word-padded data as `int[]`.
     pub contract_registry: PredicateAddress,
+    /// The address of the contract used to register programs.
+    ///
+    /// ```ignore
+    /// storage {
+    ///     programs: (b256 => Program),
+    /// }
```
[the remark this answers is not in the document]
Comment (supiket) [reply crates/node-types/src/lib.rs:69]: Oh, I missed that the `0` and `1` prefixes in contract registry were due to the index of the corresponding map in the storage block. Thanks! 

Added 0 prefix for programs.

File: crates/node/src/test_utils.rs [hunk 3]
```diff
@@ -137,21 +167,38 @@ pub fn test_invalid_block(number: Word, timestamp: Duration) -> (Block, Contract
             solutions: vec![solution],
         },
         contract,
+        program,
     )
 }
 
-pub fn test_solution(seed: Word) -> (Solution, Contract) {
-    let (solution_data, contract) = test_solution_data(seed);
+pub fn test_invalid_predicate(seed: Word) -> (Predicate, Program) {
```
[the remark this answers is not in the document]
Comment (supiket) [reply crates/node/src/test_utils.rs:174]: Agreed

File: crates/node-types/src/lib.rs (hunk 1 above)
Comment (supiket) [reply crates/node-types/src/lib.rs]: Added a comment that links to the relevant issue

**Comment by supiket** [comment]
One rename left after latest commit: 

- [ ] Rename all mentions of dec_vars in db to predicate_data

**Comment by nfurfaro** [comment]
> One rename left after latest commit:
> 
> * [ ]  Rename all mentions of dec_vars in db to predicate_data

perfect, this will unblock me nicely.

File: crates/node-db-sql/src/lib.rs [hunk 4]
```diff
@@ -109,26 +109,27 @@ pub mod table {
     }
 
     pub const BLOCK: Table = Table::new("block", create::BLOCK);
-    pub const BLOCK_SOLUTION: Table = Table::new("block_solution", create::BLOCK_SOLUTION);
-    pub const DEC_VAR: Table = Table::new("dec_var", create::DEC_VAR);
+    pub const BLOCK_SOLUTION_SET: Table =
+        Table::new("block_solution_set", create::BLOCK_SOLUTION_SET);
+    pub const PRED_DATA: Table = Table::new("pred_data", create::PRED_DATA);
```
Comment (mitchmindtree) [inline crates/node-db-sql/src/lib.rs:114]: if we want to keep these in alphabetical order it looks like we might wanna move this const too

**Review by mitchmindtree — APPROVED** [review]
Wow. HUGE work on the renaming - great job on also going though all the SQL and comments too!

This lgtm. It sounds like it's time to land these `essential-base` `develop` changes.

I don't think the `builder` update will be too major (certainly much smaller than this), so in retrospect its probably fine to open a PR for it after we land this.

I'll make the final tweaks to `essential-base/develop` now and get it ready for review and publish before merging into `main`, then we can change this to point back to crates.io :+1:

**Review by nfurfaro — APPROVED** [review]
Looks good! That's a lot of renaming and SQL.

File: crates/node-db-sql/sql/insert/block_solution_set.sql [hunk 5]
```diff
@@ -0,0 +1,8 @@
+INSERT OR IGNORE INTO
+    block_solution_set (block_id, solution_set_id, solution_set_index)
+VALUES
+    (
+        (SELECT id FROM block WHERE block.block_address = :block_address LIMIT 1),
+        (SELECT id FROM solution_set WHERE solution_set.content_hash = :solution_set_hash LIMIT 1),
```
[the remark this answers is not in the document]
Comment (supiket) [reply crates/node-db-sql/sql/insert/block_solution_set.sql:6]: Opened #167

File: Cargo.toml [hunk 6]
```diff
@@ -17,10 +17,10 @@ axum = "0.7.7"
...
 crossbeam = "0.8"
 dirs = "5"
-essential-check = { git = "https://github.com/essential-contributions/essential-base.git", branch = "main" }
-essential-hash = { git = "https://github.com/essential-contributions/essential-base.git", branch = "main" }
-essential-sign = { git = "https://github.com/essential-contributions/essential-base.git", branch = "main" }
-essential-types = { git = "https://github.com/essential-contributions/essential-base.git", branch = "main" }
+essential-check = { version = "0.10.0"}
+essential-hash = { version = "0.8.0" }
+essential-sign = { version = "0.8.0" }
+essential-types = { version = "0.6.0" }
```
Comment (mitchmindtree) [inline Cargo.toml:23]:
```suggestion
essential-check = "0.10.0"
essential-hash = "0.8.0"
essential-sign = "0.8.0"
essential-types = "0.6.0"
```

**Review by mitchmindtree — APPROVED** [review]

**Pull request merged**